sdk/go/common/resource/config/object.go

```diff
@@ -0,0 +1,487 @@
+		return elem.Delete(prefix, path[1:])
+	case map[string]object:
+		key, ok := path[0].(string)
+		if !ok {
+			return nil
+		}
+		if len(path) == 1 {
+			if len(v) == 2 {
+				keys := make([]string, 0, 2)
+				for k := range v {
+					if k != key {
+						keys = append(keys, k)
+					}
+				}
+				if len(keys) == 1 && keys[0] == "secure" {
+					if _, ok := v["secure"].value.(string); ok {
+						return errSecureReprReserved
+					}
+				}
+			}
+
+			delete(v, key)
+			return nil
+		}
+		elem, ok := v[key]
+		if !ok {
+			return nil
+		}
+		err := elem.Delete(prefix, path[1:])
+		v[key] = elem
+		return err
+	default:
+		return nil
+	}
+}
+
+func newContainer(accessor any) any {
+	switch accessor := accessor.(type) {
+	case int:
+		return make([]object, accessor+1)
+	case string:
+		return make(map[string]object)
+	default:
+		contract.Failf("unexpected accessor kind %T", accessor)
+		return nil
+	}
+}
+
+// Set sets the member value at path to new. The path to the receiver is prefix.
+func (c *object) Set(prefix, path resource.PropertyPath, new object) error {
+	if len(path) == 0 {
+		*c = new
+		return nil
+	}
+
+	switch c.value.(type) {
+	case []object, map[string]object:
+		// OK
+	case nil:
+		c.value = newContainer(path[0])
+	default:
+		// COMPAT: If this is the first level, we create a new container and overwrite the old value rather than issuing
+		// a type error.
+		if len(prefix) == 1 {
+			c.value, c.secure = newContainer(path[0]), false
+		} else {
+			switch path[0].(type) {
+			case int:
+				return fmt.Errorf("%v: expected an array", prefix)
+			case string:
+				return fmt.Errorf("%v: expected a map", prefix)
+			default:
+				contract.Failf("unreachable")
+				return nil
+			}
+		}
+	}
+
+	prefix = append(prefix, path[0])
+	switch v := c.value.(type) {
+	case []object:
+		index, ok := path[0].(int)
+		if !ok {
+			return fmt.Errorf("%v: key for an array must be an int", prefix)
+		}
+		if index < 0 || index > len(v) {
+			return fmt.Errorf("%v: array index out of range", prefix)
+		}
+		if index == len(v) {
+			v = append(v, object{})
+			c.value = v
+		}
+		elem := &v[index]
+		return elem.Set(prefix, path[1:], new)
+	case map[string]object:
+		key, ok := path[0].(string)
+		if !ok {
+			return fmt.Errorf("%v: key for a map must be a string", prefix)
+		}
+		if len(path) == 1 && len(v) == 0 && key == "secure" {
+			if _, ok := new.value.(string); ok {
+				return errSecureReprReserved
+			}
+		}
+		elem := v[key]
+		err := elem.Set(prefix, path[1:], new)
+		v[key] = elem
+		return err
+	default:
+		contract.Failf("unreachable")
+		return nil
+	}
+}
+
+// SecureValues returns the plaintext values for any secure strings contained in the receiver.
+func (c object) SecureValues(dec Decrypter) ([]string, error) {
+	switch v := c.value.(type) {
+	case []object:
+		var values []string
+		for _, v := range v {
+			vs, err := v.SecureValues(dec)
+			if err != nil {
+				return nil, err
+			}
+			values = append(values, vs...)
```

Comment:
optional: I think this will be a fair bit of slice copying since each recursive call constructs its own new slice. This could be optimized a tad bit with an inner secureValues recursive function that accepts a `*[]string` as an argument.

```go
func (c object) secureValues(dec Decrypter, secures *[]string) error {
  switch ... {
   // ...
  case string:
    if c.secure {
      // ...
      *secures = append(*secures, plaintext)
    }

```

Reply:
> optional: I think this will be a fair bit of slice copying since each recursive call constructs its own new slice.

Yep. I don't think it's going to matter, frankly, so I'm going to leave it as-is.